packages/models-library/src/models_library/api_schemas_api_server/functions.py

```diff
@@ -0,0 +1,44 @@
+from typing import Annotated, Any, Literal, TypeAlias
+
+from models_library import projects
+from pydantic import BaseModel, Field
+
+FunctionID: TypeAlias = projects.ProjectID
+
+
+class FunctionSchema(BaseModel):
+    schema_dict: dict[str, Any] | None  # JSON Schema
+
+
+class FunctionInputSchema(FunctionSchema): ...
+
+
+class FunctionOutputSchema(FunctionSchema): ...
+
+
+class Function(BaseModel):
+    uid: FunctionID | None = None
+    title: str | None = None
+    description: str | None = None
+    input_schema: FunctionInputSchema | None = None
+    output_schema: FunctionOutputSchema | None = None
+
+    # @classmethod
+    # def compose_resource_name(cls, function_key) -> api_resources.RelativeResourceName:
+    #     return api_resources.compose_resource_name("functions", function_key)
+
+
+class StudyFunction(Function):
+    function_type: Literal["study"] = "study"
+    study_url: str
+
+
+class PythonCodeFunction(Function):
+    function_type: Literal["python_code"] = "python_code"
+    code_url: str
+
+
+FunctionUnion: TypeAlias = Annotated[
+    StudyFunction | PythonCodeFunction,
+    Field(discriminator="function_type"),
+]
```

Comment:
is this: https://docs.pydantic.dev/2.0/usage/types/unions/#discriminated-unions-aka-tagged-unions
somethign that goes about what you are trying to do here?
also the url should I guess have a `AnyUrl` type or so

Reply:
Yeah, it's based on with pedro suggested.

Agreed about the url.
But ooh my, sorry, I forgot to delete this file. There is one model file, this one is outdated.